### ISSUE TITLE
Some basic form of access control

### ISSUE THREAD

**Issue by emilazy** [acme-dns/acme-dns#216]
Currently, if running an acme-dns server locally, any user can pollute the database with `curl -X POST .../register`, and if serving the API over the network then firewalling is the only access control available.

It's probably not a good idea for acme-dns to grow complex authentication mechanisms – anyone exposing it over the network with such needs should probably be running it behind a reverse proxy anyway – but it would be nice if e.g. HTTP `Authorization: Basic` with a simple shared secret was supported.

**Comment by webprofusion-chrisc** [comment]
You'd probably need it to be done in a way that your acme client doesn't need modification, so the easiest thing may be a token in the base url e.g. https://your-acme-dns/6469b532-bf2b-4523-b533-ca8ae4de3f08/register - any other compatible ways?

You can also (as a workaround) just put your http api behind a free cloudflare worker and have that require an auth token as part of the request url (your worker would then validate it, strip out the extra token and forward the real request).

**Comment by emilazy** [comment]
That seems like a reasonable approach; in my integration registration is performed separately to the ACME client so compatibility isn't a big concern, but I agree it's a relevant consideration.

I think an authentication-aware reverse proxy is the best approach for deployment over the internet, but unfortunately it doesn't work very well for the local server case; you can use network namespaces and iptables to do it, but it's a lot of moving parts. Of course, local users being able to register useless subdomains is not much of an attack vector beyond the theoretical DoS potential, but unauthenticated local TCP servers are a bugbear of mine...

In an ideal world acme-dns could listen on a Unix socket and you could use the operating system's usual access control mechanisms. Unfortunately I doubt any ACME client at all supports HTTP over Unix domain sockets :)

**Comment by jvanasco** [comment]
I address this concern with another workaround, by toggling whether the `/registration` endpoint is active or not by adjusting the `disable_registration` config option.

This method has worked for me, because I only spin up acme-dns via scripts when renewing (or enrolling) domains. The same scripts adjust the iptables too.  

I am +1 on there being some sort of access. I like the idea of using a generic token prefix, since that could be easier than HTTP Authorization for most of the clients I have seen – though this is a biased view.  (The inverse can also be true and I have no idea what the capabilities are in the total ecosystem or the top 10% of clients.)

The iptables approach i've been using is described in #67

**Comment by dereckson** [comment]
A best practice could be to listen publicly to the port 53, but only bind the HTTP endpoint to a private IP accessible by all your servers.

To have a private network across all the servers, through what the ISP offers (vRack at OVH, RPN as Online, VPC at AWS, etc.), virtual cards when you manage your VMs on a hypervisor and at worst GRE tunnels when nothing else can work is a luxury very convenient to solve this kind of problems and avoid a lot of security issues.

**Comment by candlerb** [comment]
For me, I would simply like to have an "allowed_registration" set of IP blocks, comparable to "allowfrom" for DNS updates.

That is: I don't want to disable registration entirely.  But I do want to be able to do register domains on a local trusted host, and then push out the username/password/subdomain to some remote host which is going to request the certificate.

**Comment by AvrumFeldman** [comment]
I also agree that the `/register` endpoint needs some form of authentication, especially when the acme-dns server is exposed to the internet.

In https://github.com/acme-dns/acme-dns/pull/109#issuecomment-425253213 , @joohoi suggested an optional preconfigured key passed in a request header. That is probably the cleanest approach, but it would require clients to be updated to support the new header.

Instead option would be to include a high-entropy registration token in the configured API base URL, for example:

```text
https://acme-dns.example/<registration-token>
```

Existing clients that allow the acme-dns base URL to be configured could then continue making the same `/register` and `/update` requests without any code or protocol changes. The server would simply validate the path prefix before allowing registration.

I would not consider this security through obscurity. The protection would not come from hiding the existence of the endpoint, but from requiring an unguessable pre-shared value. That is effectively the same security property as passing a token in a header; the main difference is where the token is sent.
edit: I see https://github.com/acme-dns/acme-dns/pull/109#issuecomment-425959840 even provided references to show that this is an acceptable security method.

I agree that a header is preferable, since URLs are more likely to be captured in access logs, proxy logs, and monitoring systems. However, the registration token protects a deliberately low-privilege operation. If it is exposed, the consequence is limited to unauthorized registrations and possible database growth; it does not compromise existing accounts, credentials, or DNS records. Therefore, although putting credentials in a URL is not ideal, with HTTPS, a sufficiently random token, reasonable log handling, and the ability to rotate it, the logging risk seems like an acceptable compatibility trade-off for this particular endpoint.

The goal here is also fairly narrow: preventing random unauthenticated registrations and the resulting database growth. It would not replace network restrictions, rate limiting, or stronger authentication where those are available.

This would basically combine the authentication idea from #109 with the backward-compatibility concerns raised here and in #345.

**Comment by joohoi** [comment]
While not built in to acme-dns today, this approach is already possible through reverse proxying the incoming requests. It would be fairly simple to deploy a nginx / caddy / whatnot in front of acme-dns http listener - on a custom path. 

This doesn't mean it shouldn't be implemented, but I would love to figure out the perfect implementation and if it's needed as I like to avoid adding features with no substantial value. For background reference, I've been running the public auth.acme-dns.io - instance for testing purposes for nine years or so and the database is still measured in megabytes instead of gigabytes.

**Comment by mshafer1** [comment]
I'll add that I have successfully prototyped doing mTLS with nginx https://github.com/mshafer1/mTLS-acme-dns-host
And deployed with mTLS behind Cloudflar's 0 trust tunnel https://github.com/mshafer1/acme-dns-host

**Comment by AvrumFeldman** [comment]
> While not built in to acme-dns today, this approach is already possible through reverse proxying the incoming requests. It would be fairly simple to deploy a nginx / caddy / whatnot in front of acme-dns http listener - on a custom path.

I agree that a reverse proxy can already enforce this. The benefit of native support would mainly be simpler and more consistent deployment, without requiring maintaining additional sometimes complex components.


> This doesn't mean it shouldn't be implemented, but I would love to figure out the perfect implementation and if it's needed as I like to avoid adding features with no substantial value.

Your pushback is fair and accepted and I'd like to clarify that I'm in no position to demand any features. I can explain __my__ needs and reasoning for the feature request and that might help you make a decision whether you want to implement it or accept a PR.

Also I was trying to dismiss the claims that this is security by obscurity.

> For background reference, I've been running the public auth.acme-dns.io - instance for testing purposes for nine years or so and the database is still measured in megabytes instead of gigabytes.

My concern is slightly different, though. An operator may simply not want infrastructure deployed for private organizational use to accept registrations from anyone who can reach the endpoint, even if those registrations are low privilege and cause little measurable harm. The requirement is more about controlling who is allowed to use the service than protecting against a serious security impact.

**Comment by jvanasco** [comment]
I'm glad to see some rekindled discussion on this after several years. I'd still like to see this implemented, and as a token as well (originally proposed by chris of certifytheweb).

We use a mixture of server and gateway firewalls, and front https traffic through nginx proxies ; we also use automated setup and teardown routines run and expose acme-dns on demand.  I don't like exposing knowledge of network services or topology to external actors, and I am more concerned when security related projects (like x509 provisioning) are involved. The current behavior makes it easy for simple automated tools to figure out a server is acting like acme-dns. A random token prefix would make it considerably harder.

**Comment by joohoi** [comment]
Thanks for chipping (back) in @jvanasco and amplifying the message. Helps to confirm that it's not a single-user need today regardless how well it's being justified.

**Comment by Lombra** [comment]
My vote would go to the "allowed IPs" approach. Very simple solution that probably doesn't require a lot of changes, and I feel covers the surely very common case of allowing anyone to update their own records, using the credentials that only I may provide. It is also similar to the already existing, per record `allow_from` field.

**Comment by jvanasco** [comment]
> My vote would go to the "allowed IPs" approach. 

IMHO the benefit from that work effort would be minimal, as it can already be implemented using the native os firewall; acme-dns runs each service on a port, and native os firewalls can allow/deny IPs based on ports.  it also means you need to register the future API client's IP on the server before they can use the service. An ip rules file could be managed with the acme-dns assets, and referenced by the OS; implementing this won't save any work in management or expand options.

In comparison, using a token based prefix would remove the need for having to front acme-dns with  another user-installed service, like Nginx, to enable the filtering.  

The prefix idea is really a variant of the old Apache::AuthTkt concept, where gateway servers inject AuthTkt headers onto requests as they proxy into application parts: the presence of a given secret header indicates where on the LAN the request was proxied from, and the absence means the request was erroneously or insecurely forwarded down the pipeline and should be ignored.

> Helps to confirm that it's not a single-user need today regardless how well it's being justified.

@joohoi If this might move the dial for you a bit...  the "random token prefix" could default to mounting the service under "/api/" (or "/v1/" or even "/acme-dns/" ) in the configuration file –  then a user could customize it to be whatever custom value they want.

**Comment by candlerb** [comment]
> > My vote would go to the "allowed IPs" approach.
> 
> IMHO the benefit from that work effort would be minimal, as it can already be implemented using the native os firewall; acme-dns runs each service on a port, and native os firewalls can allow/deny IPs based on ports

That doesn't work because `/register` and `/update` API calls are on the same port. What I need is to allow anyone on the Internet who has valid credentials to call `/update`, but not to allow random unauthenticated people on the Internet to call `/register`. Right now I am running with a local patch for this: https://github.com/joohoi/acme-dns/pull/388

My usage model is where I perform the registration on behalf of clients, and I issue the subdomain/username/path creds to them out-of-band.

It was mentioned above that this could be done with a proxy in front of acme-dns listening on 127.0.0.1, but I note that:

1. The proxy would need to decrypt the https session, which means it needs acme-dns's own key and certificate (or to request a duplicate certificate for the same domain name)
2. acme-dns needs to honor the `X-Forwarded-For` header for the "allowfrom" functionality to work (EDIT: I see this functionality *does* already exist, so that's not a problem)

I think this would be the approach if you have some complex requirement - e.g. you want to use OpenID Connect to authenticate the `register` endpoint.

For simple use, all we need is to restrict access to `/register`, and that could be done with IP-based limitation and/or Basic Auth and/or token in the POST body for that endpoint.  Running `/register` on a different port to the other API calls would be another way, so that it could be firewalled.

I prefer not to have an API prefix because (a) it has to be known by anyone who has `update` access, (b) whenever it is changed, everyone who has `update` access has to change it too. (A prefix just for `/register` would be OK, but I see no advantage for that over using Basic Auth or a token: clients would still need modifying).

EDIT: another thing I just thought of was to run two instances of acme-dns: one with `disable_registration` set, and the other without it but bound to a different port which is firewalled off. I'm not sure if the two instances would interfere with each other though. If that's the recommended approach then it should be documented as valid.

**Comment by bendem** [comment]
I personally don't think that reverse proxies are "complex components". They are basic building blocks that almost everybody is using in some form (be it https termination, application firewall or load balancing). Reverse proxies are the perfect place to configure authentication and acme-dns supports them natively.

I also think that using secrets in http urls is a bad solution, it spills everywhere in the logs, there is a reason we use headers for authentication.

We went the route of having acme-dns listen on localhost and the reverse proxy only proxies /update. Registration happens on the server on localhost or not at all.

Just my two cents, it's ok to implement something in acme-dns, but stating that reverse proxies are too complex so auth needs to be implemented in acme-dns itself is weird to me.

**Comment by AvrumFeldman** [comment]
> I personally don't think that reverse proxies are "complex components".

1. There is value in allowing the application to run securely enough without requiring additional infrastructure in front of it.
2. IMHO complexity is relative. A reverse proxy may be straightforward for an experienced person but not everyone is experienced with reverse proxies.
3. Not every environment already uses one, so it becomes another piece of software to set up and babysit.

> I also think that using secrets in http urls is a bad solution, it spills everywhere in the logs, there is a reason we use headers for authentication.

1. Your point is valid in general, but the concern has to be evaluated in the context of the threat model presented. This isn't a top secret password. It's just meant to prevent drive by usage, since there is no real security risk by someone registering to your instance.
2. secrets in URLs are a valid method of auth. See <sup>[[1](https://security.stackexchange.com/q/118975/29865)][[2](https://security.stackexchange.com/q/89108/29865)]</sup> brought by @Ajedi32 in https://github.com/acme-dns/acme-dns/pull/109#issuecomment-425959840

I agree that a request header would be cleaner when client compatibility is not a concern. The URL-token proposal is primarily a compatibility trade-off for existing clients that only allow the API base URL to be configured.

**Comment by jvanasco** [comment]
> EDIT: another thing I just thought of was to run two instances of acme-dns: one with disable_registration set, and the other without it but bound to a different port

I was actually about to suggest that; I use that approach a lot with other projects - including ones with sqlite3 backed databases (which might be your backend).   IIRC you should be able to do that by configuring the dns resolver on the second instance to use a port other than `:53` (there is no way to disable that - however the docstrings suggest the port can be specified).   You could then use a IP Chains to manage a file of whitelisted IPs alongside your 2 config files.  I shared some info on the IP chains method in #67.  You only need one binary, and just start each one with a different config file.  

> I personally don't think that reverse proxies are "complex components". They are basic building blocks that almost everybody is using in some form (be it https termination, application firewall or load balancing). 
> but stating that reverse proxies are too complex so auth needs to be implemented in acme-dns itself is weird to me.

A reverse proxy is a core building block – but it is another piece of software that needs to be installed, maintained and monitored on a system. The purpose of the intended integration would be to offload the application security onto it, and IMHO, offloading the security control for the application used to validate the SSL Certificate claims of an organization onto a gateway app is a (needlessly) complex integration.  Operating Systems have built-in firewalls, so leveraging that as part of the security via IP/Port rules is fine to me – it would be nice to have support for it natively, but it can be generally accomplished without installing any additional software or services.  Having to install and support a second server, for the sole purpose of securing a first server, is something that should be avoided.

One project I have uses 2 acme-dns installations in production.  One of them is on a public server with https traffic for our client's to use if needed - we toss the endpoint behind nginx for permissions, and that works perfectly fine.  The other installation is on an internal server that does not have public https traffic routed to it; nginx/apache is not installed on it, and won't be installed on it due to organization policies. 

Whether you're implementing a token prefix or a http:auth permissions header, clients will need to be updated, and changes will break.  While I'd rather see a secret header than a prefix too, a prefix should be more compatible with clients.

**Comment by dereckson** [comment]
For reference, nginx has a ngx_http_geo_module that allow to create such allowlist.

At Nasqueron, we have acme-dns behind nginx, and create the allowlist there. 

Here the minimal implementation code:

```
geo $acme_allowed_server {
    default 0;

    1.2.3.4 1;
    1.2.3.5 1;
}

map $acme_allowed_server $not_a_acme_allowed_server {
    default 0;
    0 1;
}

server {
    # ...

    location / {
        if ($not_a_nasqueron_server) {
            return 403;
        }

        proxy_pass http://.....;
  }
}
```

References:
- https://nginx.org/en/docs/http/ngx_http_geo_module.html
- https://github.com/nasqueron/operations/blob/main/roles/webserver-core/nginx/files/includes/geo_nasqueron for the allowlist
- https://github.com/nasqueron/operations/blob/main/roles/paas-docker/nginx/files/vhosts/acme_dns.conf#L51 for the nginx vhost

**Comment by candlerb** [comment]
> For reference, nginx has a ngx_http_geo_module that allow to create such allowlist.

That configuration would need extending so that it allows access to `/update` whilst blocking access to `/register`, which is what this issue is about, although that's relatively straightforward to do once you've got that far.

You still need to install nginx, and configure it to get its own certificate. It needs to use `https://` rather than `http://` that you showed in your example config to proxy to acme-dns. It needs to set X-Forwarded-For: or another header, and acme-dns needs `use_header=true`. And there must be a mechanism that blocks direct access to acme-dns from the outside world (e.g. bind it to 127.0.0.1). It's all doable, just somewhat painful.

You could equally argue the configuration setting `disable_registration = true` isn't needed in acme-dns, because you could instead put a proxy in front of it which blocks the `/register` path.  However, the acme-dns setting is a more convenient and less error-prone way to cover that use case.
